File: 0322-coin-change/0322-coin-change.py
class Solution:
    def coinChange(self, coins: List[int], amount: int) -> int:
        dp = [0] * (amount + 1)
        coins.sort()
        coinSet = set(coins)

        for i in range(1, len(dp)):
            if i in coinSet:
                dp[i] = 1
                continue
            MIN = float('inf')
            for coin in coins:
                if coin > i:
                    break
                MIN = min(MIN, 1 + dp[i - coin])
            dp[i] = MIN

        return dp[-1] if dp[-1] != float('inf') else -1

File: 0322-coin-change/0322-coin-change.cpp
class Solution {
public:
    int coinChange(vector<int>& coins, int amount) {

        vector<int> coinsRequired {0};
        unordered_set<int> coinSet (coins.begin(), coins.end());

        for (int i = 1; i <= amount; ++i) {
            if (coinSet.contains(i)) {
                coinsRequired.push_back(1);
                continue;
            }
            int minCoins = amount + 1;
            for (const int &c : coins) {
                if (i - c >= 0 && coinsRequired[i - c] != -1) {
                    minCoins = min(minCoins, coinsRequired[i - c] + 1);
                }
            }
            minCoins = minCoins <= amount ? minCoins : -1;
            coinsRequired.push_back(minCoins);
        }

        return coinsRequired.back();        
    }
};